File: sdk/typescript/src/mcp-adapter.ts
// TODO: Rewrite to use StdioClientTransport (subprocess) instead of InMemoryTransport.
// The Python SDK has already been migrated. See sdk/python/omni_connector/mcp_adapter.py.
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { InMemoryTransport } from '@modelcontextprotocol/sdk/inMemory.js';
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import type {
  ActionDefinition,
  ActionResponse,
  McpPromptDefinition,
  McpResourceDefinition,
} from './models.js';
import {
  createActionResponseSuccess,
  createActionResponseFailure,
} from './models.js';
import { getLogger } from './logger.js';

const logger = getLogger('sdk:mcp-adapter');

/**
 * Bridges an MCP McpServer into Omni's connector protocol.
 *
 * Uses InMemoryTransport to create an in-process client-server pair,
 * then interacts with the server through the standard MCP Client API.
 */
export class McpAdapter {
  private server: McpServer;
  private client: Client | null = null;

  constructor(mcpServer: McpServer) {
    this.server = mcpServer;
  }

  private async ensureConnected(): Promise<Client> {
    if (this.client) {
      return this.client;
    }
    const [clientTransport, serverTransport] = InMemoryTransport.createLinkedPair();
    await this.server.connect(serverTransport);
    this.client = new Client({ name: 'omni-mcp-adapter', version: '1.0.0' });
    await this.client.connect(clientTransport);
    return this.client;
  }

  async getActionDefinitions(): Promise<ActionDefinition[]> {
    const client = await this.ensureConnected();
    const { tools } = await client.listTools();
    const actions: ActionDefinition[] = [];

    for (const tool of tools) {
      const params: Record<string, { type: string; required: boolean; description?: string }> = {};
      const inputSchema = tool.inputSchema ?? {};
      const properties = (inputSchema as Record<string, unknown>).properties as
        | Record<string, Record<string, unknown>>
        | undefined ?? {};
      const requiredSet = new Set(
        ((inputSchema as Record<string, unknown>).required as string[] | undefined) ?? []
      );

      for (const [paramName, paramSchema] of Object.entries(properties)) {
        params[paramName] = {
          type: (paramSchema.type as string) ?? 'string',
          required: requiredSet.has(paramName),
          description: paramSchema.description as string | undefined,
        };
      }

      const isReadOnly = tool.annotations?.readOnlyHint === true;
      actions.push({
        name: tool.name,
        description: tool.description ?? '',
        parameters: params,
        mode: isReadOnly ? 'read' : 'write',
      });
    }

    return actions;
  }

  async getResourceDefinitions(): Promise<McpResourceDefinition[]> {
    const client = await this.ensureConnected();
    const definitions: McpResourceDefinition[] = [];

    const { resourceTemplates } = await client.listResourceTemplates();
    for (const tmpl of resourceTemplates) {
      definitions.push({
        uri_template: tmpl.uriTemplate,
        name: tmpl.name,
        description: tmpl.description,
        mime_type: tmpl.mimeType,
      });
    }

    const { resources } = await client.listResources();
    for (const res of resources) {
      definitions.push({
        uri_template: res.uri,
        name: res.name,
        description: res.description,
        mime_type: res.mimeType,
      });
    }

    return definitions;
  }

  async getPromptDefinitions(): Promise<McpPromptDefinition[]> {
    const client = await this.ensureConnected();
    const { prompts } = await client.listPrompts();
    return prompts.map((prompt) => ({
      name: prompt.name,
      description: prompt.description,
      arguments: (prompt.arguments ?? []).map((arg) => ({
        name: arg.name,
        description: arg.description,
        required: arg.required ?? false,
      })),
    }));
  }

  async executeTool(
    name: string,
    args: Record<string, unknown>
  ): Promise<ActionResponse> {
    const client = await this.ensureConnected();

    try {
      const result = await client.callTool({ name, arguments: args });
      if (result.isError) {
        const errorText = (result.content as Array<{ type: string; text?: string }>)
          .filter((c) => c.type === 'text')
          .map((c) => c.text ?? '')
          .join('\n');
        return createActionResponseFailure(errorText || 'Tool execution failed');
      }

      if (result.structuredContent && typeof result.structuredContent === 'object') {
        return createActionResponseSuccess(
          result.structuredContent as Record<string, unknown>
        );
      }

      const textParts: string[] = [];
      for (const block of result.content as Array<{ type: string; text?: string; mimeType?: string }>) {
        if (block.type === 'text' && block.text) {
          textParts.push(block.text);
        } else if (block.type === 'image') {
          textParts.push(`[binary: ${block.mimeType ?? 'unknown'}]`);
        }
      }
      return createActionResponseSuccess({ content: textParts.join('\n') });
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      logger.error({ err }, `MCP tool ${name} failed`);
      return createActionResponseFailure(message);
    }
  }

  async readResource(
    uri: string
  ): Promise<{ contents: Array<Record<string, unknown>> }> {
    const client = await this.ensureConnected();
    const result = await client.readResource({ uri });
    const items: Array<Record<string, unknown>> = [];

    for (const item of result.contents) {
      const entry: Record<string, unknown> = { uri: item.uri ?? uri };
      if ('text' in item && item.text != null) {
        entry.text = item.text;
      }
      if ('blob' in item && item.blob != null) {
        entry.blob = item.blob;
      }
      if (item.mimeType) {
        entry.mime_type = item.mimeType;
      }
      items.push(entry);
    }
    return { contents: items };
  }

  async getPrompt(
    name: string,
    args?: Record<string, unknown>
  ): Promise<{ description?: string; messages: Array<Record<string, unknown>> }> {
    const client = await this.ensureConnected();
    const result = await client.getPrompt({
      name,
      arguments: args as Record<string, string> | undefined,
    });
    const messages: Array<Record<string, unknown>> = [];
    for (const msg of result.messages) {
      let contentData: Record<string, unknown>;
      if (msg.content.type === 'text') {
        contentData = { type: 'text', text: msg.content.text };
      } else {
        contentData = { type: msg.content.type };
      }
      messages.push({ role: msg.role, content: contentData });
    }
    return { description: result.description, messages };
  }
}

File: sdk/python/omni_connector/mcp_adapter.py
from __future__ import annotations

import logging
from contextlib import AsyncExitStack
from typing import Any

from mcp.client.session import ClientSession
from mcp.client.stdio import StdioServerParameters, stdio_client

from .models import (
    ActionDefinition,
    ActionParameter,
    ActionResponse,
    McpPromptArgument,
    McpPromptDefinition,
    McpResourceDefinition,
)

logger = logging.getLogger(__name__)


class McpAdapter:
    """Bridges an external MCP server (subprocess) into Omni's connector protocol.

    Spawns the MCP server as a subprocess, communicates via stdio transport
    (newline-delimited JSON-RPC over stdin/stdout). Works with MCP servers
    written in any language.
    """

    def __init__(self, server_params: StdioServerParameters) -> None:
        self._base_params = server_params
        self._current_env: dict[str, str] | None = None
        self._session: ClientSession | None = None
        self._exit_stack: AsyncExitStack | None = None

    async def ensure_connected(
        self, env: dict[str, str] | None = None
    ) -> ClientSession:
        """Start or reuse the MCP subprocess.

        If *env* differs from the current subprocess environment, the old
        process is torn down and a new one is spawned with the updated env.
        """
        if self._session is not None and env == self._current_env:
            return self._session

        await self.disconnect()

        params = StdioServerParameters(
            command=self._base_params.command,
            args=self._base_params.args,
            env={**(self._base_params.env or {}), **(env or {})},
            cwd=self._base_params.cwd,
        )

        self._exit_stack = AsyncExitStack()
        try:
            read_stream, write_stream = await self._exit_stack.enter_async_context(
                stdio_client(params)
            )
            session = await self._exit_stack.enter_async_context(
                ClientSession(read_stream, write_stream)
            )
            await session.initialize()
            self._session = session
            self._current_env = env
            logger.info("MCP subprocess connected: %s", self._base_params.command)
            return session
        except Exception:
            await self._exit_stack.aclose()
            self._exit_stack = None
            raise

    async def disconnect(self) -> None:
        """Terminate the MCP subprocess if running."""
        if self._exit_stack is not None:
            try:
                await self._exit_stack.aclose()
            except Exception:
                logger.warning("Error closing MCP subprocess", exc_info=True)
            self._exit_stack = None
        self._session = None
        self._current_env = None

    async def get_action_definitions(
        self, env: dict[str, str] | None = None
    ) -> list[ActionDefinition]:
        session = await self.ensure_connected(env)
        result = await session.list_tools()
        actions: list[ActionDefinition] = []
        for tool in result.tools:
            params: dict[str, ActionParameter] = {}
            input_schema = tool.inputSchema or {}
            properties = input_schema.get("properties", {})
            required_set = set(input_schema.get("required", []))
            for param_name, param_schema in properties.items():
                params[param_name] = ActionParameter(
                    type=param_schema.get("type", "string"),
                    required=param_name in required_set,
                    description=param_schema.get("description"),
                )
            is_read_only = bool(tool.annotations and tool.annotations.readOnlyHint)
            actions.append(
                ActionDefinition(
                    name=tool.name,
                    description=tool.description or "",
                    parameters=params,
                    mode="read" if is_read_only else "write",
                )
            )
        return actions

    async def get_resource_definitions(
        self, env: dict[str, str] | None = None
    ) -> list[McpResourceDefinition]:
        session = await self.ensure_connected(env)
        definitions: list[McpResourceDefinition] = []

        templates_result = await session.list_resource_templates()
        for tmpl in templates_result.resourceTemplates:
            definitions.append(
                McpResourceDefinition(
                    uri_template=str(tmpl.uriTemplate),
                    name=tmpl.name,
                    description=tmpl.description,
                    mime_type=tmpl.mimeType,
                )
            )

        resources_result = await session.list_resources()
        for res in resources_result.resources:
            definitions.append(
                McpResourceDefinition(
                    uri_template=str(res.uri),
                    name=res.name,
                    description=res.description,
                    mime_type=res.mimeType,
                )
            )

        return definitions

    async def get_prompt_definitions(
        self, env: dict[str, str] | None = None
    ) -> list[McpPromptDefinition]:
        session = await self.ensure_connected(env)
        result = await session.list_prompts()
        definitions: list[McpPromptDefinition] = []
        for prompt in result.prompts:
            args = [
                McpPromptArgument(
                    name=arg.name,
                    description=arg.description,
                    required=arg.required or False,
                )
                for arg in (prompt.arguments or [])
            ]
            definitions.append(
                McpPromptDefinition(
                    name=prompt.name,
                    description=prompt.description,
                    arguments=args,
                )
            )
        return definitions

    async def execute_tool(
        self, name: str, arguments: dict[str, Any], env: dict[str, str] | None = None
    ) -> ActionResponse:
        try:
            session = await self.ensure_connected(env)
            result = await session.call_tool(name, arguments)
            text_parts: list[str] = []
            for block in result.content:
                if hasattr(block, "text"):
                    text_parts.append(block.text)
                elif hasattr(block, "data"):
                    text_parts.append(
                        f"[binary: {getattr(block, 'mimeType', 'unknown')}]"
                    )
            content = "\n".join(text_parts)
            if result.isError:
                return ActionResponse.failure(content)
            return ActionResponse.success({"content": content})
        except Exception as e:
            logger.error("MCP tool %s failed: %s", name, e)
            return ActionResponse.failure(str(e))

    async def read_resource(
        self, uri: str, env: dict[str, str] | None = None
    ) -> dict[str, Any]:
        session = await self.ensure_connected(env)
        result = await session.read_resource(uri)
        items: list[dict[str, Any]] = []
        for item in result.contents:
            entry: dict[str, Any] = {"uri": str(item.uri)}
            if hasattr(item, "text") and item.text is not None:
                entry["text"] = item.text
            if hasattr(item, "mimeType") and item.mimeType:
                entry["mime_type"] = item.mimeType
            items.append(entry)
        return {"contents": items}

    async def get_prompt(
        self,
        name: str,
        arguments: dict[str, Any] | None = None,
        env: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        session = await self.ensure_connected(env)
        result = await session.get_prompt(name, arguments)
        messages: list[dict[str, Any]] = []
        for msg in result.messages:
            content_data: dict[str, Any]
            if hasattr(msg.content, "text"):
                content_data = {"type": "text", "text": msg.content.text}
            else:
                content_data = {"type": "unknown"}
            messages.append({"role": msg.role, "content": content_data})
        return {"description": result.description, "messages": messages}
